In [1]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [8]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
lending_df = pd.read_csv(Path('Resources/lending_data.csv'))

# Review the DataFrame
lending_df.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0

### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [5]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = lending_df['loan_status']

# Separate the X variable, the features
X = lending_df.drop(columns='loan_status')

In [6]:
# Review the y variable Series
y[:5]

0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64

In [9]:
# Review the X variable DataFrame
X.head()

loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000

### Step 3: Split the data into training and testing datasets by using `train_test_split`.

In [10]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [19]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_regression_model = LogisticRegression(solver='lbfgs', random_state=1, max_iter=300)

# Fit the model using training data
lr_model = logistic_regression_model.fit(X_train, y_train)


### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [23]:
# Make a prediction using the testing data
 #Generate training predictions
training_predictions = lr_model.predict(X_train)

#Generate testing predictions
testing_predictions = logistic_regression_model.predict(X_test)
testing_predictions 


array([0, 0, 0, ..., 0, 0, 0])

### Step 3: Evaluate the model’s performance by doing the following:

* Generate a confusion matrix.

* Print the classification report.

In [28]:
# Generate a confusion matrix for the model
# Create and save the confusion matrix for the training data
training_matrix = confusion_matrix(y_train, training_predictions)
training_matrix_df = pd.DataFrame(training_matrix, index=['Actual Negative', 'Actual Positive'], columns=['Predicted Negative', 'Predicted Positive'])

# Print the confusion matrix for the training data
print("Training Confusion Matrix:")
print(training_matrix_df)

Training Confusion Matrix:
                 Predicted Negative  Predicted Positive
Actual Negative               55988                 283
Actual Positive                 137                1744


In [27]:
# Create and save the confusion matrix for the testing data
testing_matrix = confusion_matrix(y_test, testing_predictions)
testing_matrix_df = pd.DataFrame(testing_matrix, index=['Actual Negative', 'Actual Positive'], columns=['Predicted Negative', 'Predicted Positive'])

# Print the confusion matrix for the testing data
print("Testing Confusion Matrix:")
print(testing_matrix_df)

Testing Confusion Matrix:
                 Predicted Negative  Predicted Positive
Actual Negative               18658                 107
Actual Positive                  37                 582


In [31]:
# Print the classification report for the model training data
training_report = classification_report(y_train, training_predictions, target_names=['Healthy Loan', 'High-Risk Loan'])

# Print the training classification report
print(training_report)

                precision    recall  f1-score   support

  Healthy Loan       1.00      0.99      1.00     56271
High-Risk Loan       0.86      0.93      0.89      1881

      accuracy                           0.99     58152
     macro avg       0.93      0.96      0.94     58152
  weighted avg       0.99      0.99      0.99     58152



In [32]:
# Print the classification report for the model testing data
testing_report = classification_report(y_test, testing_predictions, target_names=['Healthy Loan', 'High-Risk Loan'])

# Print the testing classification report
print(testing_report)

                precision    recall  f1-score   support

  Healthy Loan       1.00      0.99      1.00     18765
High-Risk Loan       0.84      0.94      0.89       619

      accuracy                           0.99     19384
     macro avg       0.92      0.97      0.94     19384
  weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** Based on the report, the logistic regression model performs quite well in predicting both the **Healthy Loan** and **High-Risk Loan** labels, but there are some differences in performance between the two classes:

1. **Healthy Loan**:
   - **Precision**: 1.00 — The model is nearly perfect at identifying healthy loans (very few false positives).
   - **Recall**: 0.99 — Almost all healthy loans are correctly identified (very few false negatives).
   - **F1-score**: 1.00 — This high F1-score reflects the strong balance between precision and recall for this class.
   
   Overall, the model is extremely accurate in predicting healthy loans.

2. **High-Risk Loan**:
   - **Precision**: 0.84 — There are more false positives for high-risk loans compared to healthy loans, indicating that some healthy loans are misclassified as high-risk.
   - **Recall**: 0.94 — The model captures the majority of high-risk loans (most true positives are identified).
   - **F1-score**: 0.89 — While high, the F1-score is lower than that of healthy loans due to the lower precision.

   Although the model does well in predicting high-risk loans, it’s less accurate than it is with healthy loans. The lower precision for high-risk loans indicates that it sometimes mistakenly flags healthy loans as high-risk.

3. **Overall Accuracy**:
   - The model has a very high accuracy of **0.99**, meaning it correctly classifies 99% of all loans, which is heavily influenced by the large number of healthy loans.
   
4. **Macro and Weighted Averages**:
   - The **macro average** (unweighted) F1-score is 0.94, indicating strong performance across both classes.
   - The **weighted average** F1-score is 0.99, reflecting the dominance of healthy loans in the dataset.

### Summary
The logistic regression model is highly effective at predicting **healthy loans** but slightly less effective with **high-risk loans** due to lower precision. This means it occasionally misclassifies some healthy loans as high-risk, but it still identifies the majority of high-risk loans correctly.

---